File: sdks/php/tests/Unit/LoginTest.php
<?php

declare(strict_types=1);

namespace Hearth\Tests\Unit;

use GuzzleHttp\Psr7\HttpFactory;
use GuzzleHttp\Psr7\Response;
use Hearth\HearthClient;
use Hearth\Types\LoginBeginResult;
use Hearth\Types\TokenResponse;
use PHPUnit\Framework\TestCase;
use Psr\Http\Client\ClientInterface;
use Psr\Http\Message\RequestInterface;
use Psr\Http\Message\ResponseInterface;

/**
 * Unit tests for HearthClient::beginLogin() and completeLogin() (HEA-1592).
 */
final class LoginTest extends TestCase
{
    // -------------------------------------------------------------------------
    // Helpers
    // -------------------------------------------------------------------------

    private function makeClient(array $responses, array $opts = []): HearthClient
    {
        $mock    = new LoginSequentialMockClient($responses);
        $factory = new HttpFactory();

        return new HearthClient(
            issuerUrl:      $opts['issuerUrl']    ?? 'https://auth.example.com',
            clientId:       $opts['clientId']     ?? 'test-client',
            clientSecret:   $opts['clientSecret'] ?? 's3cr3t',
            httpClient:     $mock,
            requestFactory: $factory,
            streamFactory:  $factory,
        );
    }

    private function discoveryResponse(): Response
    {
        return new Response(200, ['Content-Type' => 'application/json'], json_encode([
            'issuer'                 => 'https://auth.example.com',
            'authorization_endpoint' => 'https://auth.example.com/authorize',
            'token_endpoint'         => 'https://auth.example.com/token',
            'jwks_uri'               => 'https://auth.example.com/.well-known/jwks.json',
        ]));
    }

    private function tokenResponseData(): array
    {
        return ['access_token' => 'eyJ.tok.en', 'token_type' => 'Bearer', 'expires_in' => 3600];
    }

    // -------------------------------------------------------------------------
    // beginLogin
    // -------------------------------------------------------------------------

    public function testBeginLoginReturnsLoginBeginResult(): void
    {
        $client = $this->makeClient([$this->discoveryResponse()]);
        $result = $client->beginLogin('https://app.example.com/callback');

        self::assertInstanceOf(LoginBeginResult::class, $result);
        self::assertNotEmpty($result->authorizationUrl);
        self::assertNotEmpty($result->state);
        self::assertNotEmpty($result->codeVerifier);
    }

    public function testBeginLoginCodeChallengeMatchesVerifier(): void
    {
        $client = $this->makeClient([$this->discoveryResponse()]);
        $result = $client->beginLogin('https://app.example.com/callback');

        parse_str(parse_url($result->authorizationUrl, PHP_URL_QUERY), $params);
        $challenge = $params['code_challenge'] ?? '';

        $expected = rtrim(strtr(base64_encode(hash('sha256', $result->codeVerifier, true)), '+/', '-_'), '=');
        self::assertSame($expected, $challenge, 'code_challenge must be BASE64URL(SHA256(codeVerifier))');
    }

    public function testBeginLoginStateAppearsInUrl(): void
    {
        $client = $this->makeClient([$this->discoveryResponse()]);
        $result = $client->beginLogin('https://app.example.com/callback');

        parse_str(parse_url($result->authorizationUrl, PHP_URL_QUERY), $params);
        self::assertSame($result->state, $params['state'] ?? '', 'state in URL must match returned state');
    }

    public function testBeginLoginUrlContainsRequiredParams(): void
    {
        $client = $this->makeClient([$this->discoveryResponse()]);
        $result = $client->beginLogin('https://app.example.com/callback', 'openid profile');

        parse_str(parse_url($result->authorizationUrl, PHP_URL_QUERY), $params);
        self::assertSame('code', $params['response_type'] ?? '');
        self::assertSame('test-client', $params['client_id'] ?? '');
        self::assertSame('https://app.example.com/callback', $params['redirect_uri'] ?? '');
        self::assertSame('openid profile', $params['scope'] ?? '');
        self::assertSame('S256', $params['code_challenge_method'] ?? '');
    }

    public function testBeginLoginDefaultsScopeToOpenid(): void
    {
        $client = $this->makeClient([$this->discoveryResponse()]);
        $result = $client->beginLogin('https://app.example.com/callback');

        parse_str(parse_url($result->authorizationUrl, PHP_URL_QUERY), $params);
        self::assertSame('openid', $params['scope'] ?? '');
    }

    // -------------------------------------------------------------------------
    // completeLogin
    // -------------------------------------------------------------------------

    public function testCompleteLoginReturnsTokenResponse(): void
    {
        $client = $this->makeClient([
            $this->discoveryResponse(),
            new Response(200, ['Content-Type' => 'application/json'], json_encode($this->tokenResponseData())),
        ]);

        $result = $client->completeLogin('auth-code-xyz', 'my-verifier-abc', 'https://app.example.com/callback');

        self::assertInstanceOf(TokenResponse::class, $result);
        self::assertSame('eyJ.tok.en', $result->accessToken);
    }

    public function testCompleteLoginSendsCodeVerifierToTokenEndpoint(): void
    {
        $mock    = new LoginCapturingMockClient([
            $this->discoveryResponse(),
            new Response(200, ['Content-Type' => 'application/json'], json_encode($this->tokenResponseData())),
        ]);
        $factory = new HttpFactory();
        $client  = new HearthClient(
            issuerUrl:      'https://auth.example.com',
            clientId:       'test-client',
            clientSecret:   's3cr3t',
            httpClient:     $mock,
            requestFactory: $factory,
            streamFactory:  $factory,
        );

        $client->completeLogin('auth-code-xyz', 'my-verifier-abc', 'https://app.example.com/callback');

        // The second request is to the token endpoint
        $tokenRequest = $mock->requests[1] ?? $mock->requests[0];
        $body         = (string) $tokenRequest->getBody();
        parse_str($body, $params);

        self::assertSame('my-verifier-abc', $params['code_verifier'] ?? '', 'code_verifier must be sent to token endpoint');
        self::assertSame('auth-code-xyz', $params['code'] ?? '', 'code must be sent to token endpoint');
        self::assertSame('authorization_code', $params['grant_type'] ?? '');
    }
}

// ── Test-local helpers ────────────────────────────────────────────────────────

final class LoginSequentialMockClient implements ClientInterface
{
    private int $index = 0;

    /** @param ResponseInterface[] $responses */
    public function __construct(private readonly array $responses) {}

    public function sendRequest(RequestInterface $request): ResponseInterface
    {
        return $this->responses[$this->index++]
            ?? throw new \LogicException("No mock response for request #{$this->index}");
    }
}

final class LoginCapturingMockClient implements ClientInterface
{
    private int $index = 0;
    /** @var RequestInterface[] */
    public array $requests = [];

    /** @param ResponseInterface[] $responses */
    public function __construct(private readonly array $responses) {}

    public function sendRequest(RequestInterface $request): ResponseInterface
    {
        $this->requests[] = $request;
        return $this->responses[$this->index++]
            ?? throw new \LogicException("No mock response for request #{$this->index}");
    }
}

File: sdks/kotlin/hearth-core/src/test/kotlin/io/hearth/sdk/LoginTest.kt
package io.hearth.sdk

import kotlinx.coroutines.test.runTest
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import java.net.URLDecoder
import java.security.MessageDigest
import java.util.Base64
import kotlin.test.AfterTest
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

/**
 * Tests for HearthClient.beginLogin() and completeLogin() (HEA-1592).
 */
class LoginTest {

    private lateinit var server: MockWebServer

    @BeforeTest
    fun setUp() {
        server = MockWebServer()
        server.start()
    }

    @AfterTest
    fun tearDown() {
        server.shutdown()
    }

    private val discoveryDoc get() = """
        {
          "issuer": "${server.url("/").toString().trimEnd('/')}",
          "jwks_uri": "${server.url("/jwks")}",
          "token_endpoint": "${server.url("/token")}",
          "authorization_endpoint": "${server.url("/authorize")}",
          "userinfo_endpoint": "${server.url("/userinfo")}"
        }
    """.trimIndent()

    private fun makeClient(opts: Map<String, String?> = emptyMap()): HearthClient =
        HearthClient(
            issuerUrl = server.url("/").toString().trimEnd('/'),
            clientId = opts["clientId"] ?: "test-client",
            clientSecret = opts["clientSecret"] ?: "s3cr3t",
        )

    private fun parseQuery(query: String): Map<String, String> =
        query.split("&").associate { part ->
            val (k, v) = part.split("=", limit = 2)
            URLDecoder.decode(k, "UTF-8") to URLDecoder.decode(v, "UTF-8")
        }

    // ── beginLogin ────────────────────────────────────────────────────────────

    @Test
    fun `beginLogin returns authorizationUrl with code_challenge derived from codeVerifier`() = runTest {
        server.enqueue(MockResponse().setBody(discoveryDoc).setResponseCode(200))

        val client = makeClient()
        val result = client.beginLogin("https://app.example.com/callback")

        val url = java.net.URL(result.authorizationUrl)
        val params = parseQuery(url.query)
        val challenge = params["code_challenge"] ?: error("code_challenge missing")

        val digest = MessageDigest.getInstance("SHA-256").digest(result.codeVerifier.toByteArray())
        val expected = Base64.getUrlEncoder().withoutPadding().encodeToString(digest)
        assertEquals(expected, challenge, "code_challenge must be BASE64URL(SHA256(codeVerifier))")
    }

    @Test
    fun `beginLogin state is non-empty and present in URL`() = runTest {
        server.enqueue(MockResponse().setBody(discoveryDoc).setResponseCode(200))

        val client = makeClient()
        val result = client.beginLogin("https://app.example.com/callback")

        assertTrue(result.state.isNotEmpty(), "state must not be empty")
        val url = java.net.URL(result.authorizationUrl)
        val params = parseQuery(url.query)
        assertEquals(result.state, params["state"], "state in URL must match returned state")
    }

    @Test
    fun `beginLogin URL contains required OAuth and PKCE params`() = runTest {
        server.enqueue(MockResponse().setBody(discoveryDoc).setResponseCode(200))

        val client = makeClient()
        val result = client.beginLogin("https://app.example.com/callback", "openid profile")

        val url = java.net.URL(result.authorizationUrl)
        val params = parseQuery(url.query)
        assertEquals("code", params["response_type"])
        assertEquals("test-client", params["client_id"])
        assertEquals("https://app.example.com/callback", params["redirect_uri"])
        assertEquals("openid profile", params["scope"])
        assertEquals("S256", params["code_challenge_method"])
        assertNotNull(params["code_challenge"])
    }

    @Test
    fun `beginLogin defaults scope to openid`() = runTest {
        server.enqueue(MockResponse().setBody(discoveryDoc).setResponseCode(200))

        val client = makeClient()
        val result = client.beginLogin("https://app.example.com/callback")

        val params = parseQuery(java.net.URL(result.authorizationUrl).query)
        assertEquals("openid", params["scope"])
    }

    @Test
    fun `beginLogin throws ConfigurationError when clientId is not set`() = runTest {
        val client = HearthClient(
            issuerUrl = server.url("/").toString().trimEnd('/'),
            clientId = null,
        )
        assertFailsWith<ConfigurationError> {
            client.beginLogin("https://app.example.com/callback")
        }
    }

    // ── completeLogin ─────────────────────────────────────────────────────────

    @Test
    fun `completeLogin posts code_verifier to token endpoint`() = runTest {
        server.enqueue(MockResponse().setBody(discoveryDoc).setResponseCode(200))
        server.enqueue(
            MockResponse().setBody(
                """{"access_token":"eyJ...","token_type":"Bearer","expires_in":3600}"""
            ).setResponseCode(200)
        )

        val client = makeClient()
        val tokens = client.completeLogin(
            code = "auth-code-xyz",
            codeVerifier = "my-verifier-abc",
            redirectUri = "https://app.example.com/callback",
        )

        assertEquals("eyJ...", tokens.accessToken)

        // The second request must be to the token endpoint with the verifier.
        server.takeRequest() // discovery
        val tokenReq = server.takeRequest()
        val body = tokenReq.body.readUtf8()
        // Token endpoint receives JSON (Kotlin SDK uses JSON encoding)
        assertTrue(body.contains("my-verifier-abc"), "code_verifier missing: $body")
        assertTrue(body.contains("auth-code-xyz"), "code missing: $body")
        assertTrue(body.contains("authorization_code"), "grant_type missing: $body")
    }
}
